### ISSUE TITLE
llvm.x86.addcarry.64 intrinsic not yet implemented

### ISSUE THREAD

**Issue by shamatar** [bjorn3/rustc_codegen_cranelift#1177]
I was trying to use a cranelift backend for faster compilation times during the development and followed instructions to build a backed. Unfortunately when I try to compile e.g. this lib https://github.com/matter-labs/pairing it fails with no constructive error message (I've deleted various pathes for simplicity)

```
process didn't exit successfully: `cg_clif --crate-name pairing_ce --edition=2018 src/lib.rs --error-format=json --json=diagnostic-rendered-ansi --crate-type lib --emit=dep-info,metadata,link -C opt-level=3 -C embed-bitcode=no --cfg 'feature="default"' -C metadata=ba4f3e83328c5404 -C extra-filename=-ba4f3e83328c5404 --out-dir ... -L dependency=... --extern byteorder=... --extern cfg_if=... --extern ff=... --extern rand=... -Clink-arg=-undefined -Clink-arg=dynamic_lookup` (signal: 4, SIGILL: illegal instruction)
```

I see that `-Clink-arg=-undefined` is potentially troublesome, but there are no hints on the cause of the panic. LLVM backend build a package without any trouble.

Both debug and release modes behave the same. Dependencies of this library can use assembly, but those are under feature flags and are not enabled by default

**Comment by shamatar** [comment]
Continued checking with running `./test.sh` and found the following
```
[AOT] arbitrary_self_types_pointers_and_wrappers
[arbitrary_self_types_pointers_and_wrappers: codegen mono items] start
[arbitrary_self_types_pointers_and_wrappers: codegen mono items] end time: 39.990084ms
[AOT] alloc_system
[alloc_system                  : codegen mono items] start
[alloc_system                  : codegen mono items] end time: 37.795616ms
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: target/out/liballoc_system.rlib(lib.rmeta) has no symbols
[AOT] alloc_example
[alloc_example                 : codegen mono items] start
[alloc_example                 : codegen mono items] end time: 30.562747ms
Hello World!
[JIT] std_example
[std_example                   : codegen mono items] start
thread 'rustc' panicked at 'called `Result::unwrap()` on an `Err` value: TryFromIntError(())', /Users/alexvlasov/.cargo/git/checkouts/wasmtime-41807828cb3a7a7e/3d56728/cranelift/jit/src/backend.rs:253:82
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.54.0-nightly (e4a603270 2021-06-07) running on x86_64-apple-darwin

note: compiler flags: -C link-arg=-undefined -C link-arg=dynamic_lookup -C debuginfo=2 -C llvm-args=mode=jit -C prefer-dynamic

query stack during panic:
end of query stack
Instance { def: Item(WithOptConstParam { did: DefId(1:8893 ~ std[11f2]::sys_common::remutex::{impl#8}::drop), const_param_did: None }), substs: [ReErased, std::cell::RefCell<std::io::stdio::StderrRaw>] } _ZN96_$LT$std..sys_common..remutex..ReentrantMutexGuard$LT$T$GT$$u20$as$u20$core..ops..drop..Drop$GT$4drop17he3167ede49be3ba9E
```

**Comment by bjorn3** [comment]
By the way are you on an M1 mac with rosetta 2 or an x86_64 mac? That panic in the test suite seems to suggest that the data and code allocations by the jit are more than 2GB apart. Most of the time they are directly next to each other, but maybe rosetta 2 doesn't handle this?

Can you try to remove https://github.com/bjorn3/rustc_codegen_cranelift/blob/75eff64977ff149ead73bc0077bac5872d599aa2/scripts/tests.sh#L44-L52 and see if the rest of the test suite passes?

Can you get a backtrace for the crashing cg_clif invocation for https://github.com/matter-labs/pairing?

**Comment by bjorn3** [comment]
> I see that -Clink-arg=-undefined is potentially troublesome

This is https://github.com/bjorn3/rustc_codegen_cranelift/blob/75eff64977ff149ead73bc0077bac5872d599aa2/scripts/config.sh#L21-L24 It should probably be moved to `ext_config.sh` which doesn't apply to user code compiled with cg_clif.

**Comment by shamatar** [comment]
I do use Mac, but it's Intel based. Let me try the suggestions

**Comment by shamatar** [comment]
> Can you get a backtrace for the crashing cg_clif invocation for https://github.com/matter-labs/pairing?

Can you point me on the corresponding manual page (quick googling didn't give me an answer). Should I recompile the cranelift with debug info?

**Comment by shamatar** [comment]
> By the way are you on an M1 mac with rosetta 2 or an x86_64 mac? That panic in the test suite seems to suggest that the data and code allocations by the jit are more than 2GB apart. Most of the time they are directly next to each other, but maybe rosetta 2 doesn't handle this?
> 
> Can you try to remove
> 
> https://github.com/bjorn3/rustc_codegen_cranelift/blob/75eff64977ff149ead73bc0077bac5872d599aa2/scripts/tests.sh#L44-L52
> 
> and see if the rest of the test suite passes?
> Can you get a backtrace for the crashing cg_clif invocation for https://github.com/matter-labs/pairing?

With commented out lines test went to the benchmarks successfully on X86-64 Mac

**Comment by bjorn3** [comment]
> Can you point me on the corresponding manual page (quick googling didn't give me an answer). Should I recompile the cranelift with debug info? An information that I get for pairing library is in the first message, it's everything that I see in the terminal

That would be something like:

```
$ lldb /path/to/cg_clif --crate-name pairing_ce --edition=2018 src/lib.rs --error-format=json --json=diagnostic-rendered-ansi --crate-type lib [...]
(lldb) run
(lldb) bt all
```

That is `lldb` followed by the full commandline listed after `process didn't exit successfully` and with `cg_clif` turned into the full path to the `cg_clif` executable. (it can be found in the `build/bin` dir of your local clone of cg_clif) The backtrace may be enough with just the symbol names, but without debuginfo. You could also try enabling the debuginfo by adding the following to cg_clif's `Cargo.toml`:

```toml
[profile.release]
debug = true
```

and then recompiling cg_clif.

**Comment by bjorn3** [comment]
I tried it locally. This is what I get in a debugger:

```
trap at Instance { def: Item(WithOptConstParam { did: DefId(1:20497 ~ core[2329]::core_arch::x86_64::adx::_addcarry_u64), const_param_did: None }), substs: [] } (_ZN4core9core_arch6x86_643adx13_addcarry_u6417hf7a60e4866590308E): llvm.x86.addcarry.64
```

Based on the backtrace it is used by the ff_ce_derive proc macro when it uses num_bigint crate to parse a bigint. Only a small subset of all llvm intrinsics is emulated by cg_clif. This one isn't yet.

**Comment by shamatar** [comment]
That's interesting, I thought `adx` is never issued by rustc and LLVM at all, and only old style `add/adc` are used

**Comment by bjorn3** [comment]
num_bigint calls [`core::arch::x86_64::_addcarry_u64`](https://doc.rust-lang.org/stable/core/arch/x86_64/fn._addcarryx_u64.html) which is implemented using the `llvm.x86.addcarry.64` llvm intrinsic internally.

**Comment by shamatar** [comment]
Ok, I was confused by adx prefix in a path. Since it's most likely a stopper at the moment can I somehow help to add support of this intrinsic? Later on it will also happen in the derived code too that does addition with carry

**Comment by bjorn3** [comment]
I am rather busy today, but I can give you a bit of guidance on how to implement it: The right location to implement it would be in https://github.com/bjorn3/rustc_codegen_cranelift/blob/master/src/intrinsics/llvm.rs The function signature of the llvm intrinsic is:

```rust
#[link_name = "llvm.x86.addcarry.64"]
fn llvm_addcarry_u64(a: u8, b: u64, c: u64) -> (u8, u64);
```

The actual `_addcarry_u64` function is defined as follows:

```rust
pub unsafe fn _addcarry_u64(c_in: u8, a: u64, b: u64, out: &mut u64) -> u8 {
    let (a, b) = llvm_addcarry_u64(c_in, a, b);
    *out = b;
    a
}
```

You can construct a value of type `(u8, u64)` using `CValue::scalar_pair(a, b, layout)`. Other than that you can take a look at how other intrinsics are implemented in eg https://github.com/bjorn3/rustc_codegen_cranelift/blob/master/src/intrinsics/mod.rs (note: this is a different file. this is the list of most rust intrinsics. `llvm.rs` is just for llvm intrinsics)

**Comment by bjorn3** [comment]
Fixed in #1178.
